Sources/GRPCCore/Documentation.docc/Articles/Migration-guide.md

```diff
@@ -0,0 +1,501 @@
+runtime components for 2.x.
+
+Next let's update the service protocol that your type conforms to. In 2.x each
+service has three protocols generated for it, each offering a different level of
+granularity. You can read more about each version in the [gRPC Swift Protobuf
+documentation][2]. The variant most like 1.x is the `SimpleServiceProtocol`.
+However, it doesn't allow you access metadata. If you need access to metadata
+skip to the section called `ServiceProtocol`.
+
+### SimpleServiceProtocol
+
+The requirements for each methods are also slightly different; in 2.x the context
+type is called `ServerContext` as opposed to `GRPCAsyncServerCallContext` in 1.x.
+It also has different functionality but that will be covered later. The types
+for streaming requests and responses are also different:
+- `GRPCAsyncRequestStream<T>` became `RPCAsyncSequence<T, any Error>`, and
+- `GRPCAsyncResponseStreamWriter<T>` became `RPCWriter<T>`.
+
+The `v1_to_v2` script has a subcommand to apply all of these transformations to
+an input file. Run it now. Here's an example invocation:
+
+```sh
+./v1_to_v2 patch-service Sources/Server/Service.swift
+```
+
+If the service was contained to that file then that might be the extent of
+changes you need to make for that service. However, it's likely that types leak
+into other files. If that's the case you should continue applying these
+transformations until your app compiles again. You'll also need to stop
+passing this service to your 1.x server.
+
+Once you've gotten to a point where the package builds, commit your changes.
+Repeat this until you've done all services in your package.
+
+### ServiceProtocol
+
+> If the `SimpleServiceProtocol` worked then you can skip this section.
+
+If you're reading this section then you're likely relying on metadata in your
+service. This means you need to implement the `ServiceProtocol` instead of the
+`SimpleServiceProtocol` and the transformations you need to apply are
+aren't well suited automation. The best approach is to conform your
+service to the 1.x protocol and the 2.x protocol. Add conformance to the
+`{Service}.ServiceProtocol` where `{Service}` is the namespaced name of your
+service (if your service is called `Baz` and declared in the `foo.bar` Protocol
+Buffers package then this would be `Foo_Bar_Baz.ServiceProtocol`).
+
+Let Xcode generate stubs for the methods which haven't been implemented yet and
+fill each one with a `fatalError` so that you app builds. Each method
+should take a `ServerRequest` or `StreamingServerRequest` and context as input
+and return a `ServerResponse` or `StreamingServerResponse`. Request metadata is
+available on the request object. For single responses you can set initial and
+trailing metadata when you create the response. For streaming responses you can
+set initial metadata in the initializer and return trailing metadata from the
+closure you provide to the initializer.
```

Comment:
Should we link to the docs or tutorials in case this compressed version isn't enough to understand what's going on?